addons/bus/models/ir_async.py

```diff
@@ -0,0 +1,88 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+import logging
+from odoo import models, fields
+
+_logger = logging.getLogger(__name__)
+
+
+class IrAsync(models.Model):
+    _inherit = "ir.async"
+
+    notify = fields.Boolean(default=False)
+
+    def call_notify(self, description, method, *args, **kwargs):
+        """
+        partial-like API to create asynchronous jobs with web notification.
+
+        The method will be called later as ``target(*args, **kwargs)`` in a
+        new transaction using a copy of the current environment (user, context,
+        recordset). Arguments must be serializable in JSON.
+
+        The notifications are sent via the longpolling bus, they are:
+
+        - created, the task has been enqueued
+        - processing, the worker begins to process the task
+        - succeeded, the task succeeded with a result the user must process
+        - failed, when the task failed with an error
+        - done, the task succeeded without result or the user processed it
+
+        See the ``async_job`` javascript service.
+
+        The ``description`` is used to list the task in the systray menu
+        in the frontend. You may pass a falsy value to hide the task
+        from the systray but still get notified.
+        """
+        job = self.with_context(
+            default_name=description, default_notify=True
+        ).call(method, *args, **kwargs)
+
+        # Notify the user a task has been created
+        channel = (self._cr.dbname, 'res.partner', self.env.user.partner_id.id)
+        self.env['bus.bus'].sendone(channel, {
+            'type': 'ir.async',
+            'id': job.id,
+            'name': job.name,
+            'state': job.state,
+        })
+
+        return job
+
+    def _pre_process(self, job):
+        # Notify the user a task is processing
+        super()._pre_process(job)
+        if job['notify']:
+            channel = (self._cr.dbname, 'res.partner', self.env.user.partner_id.id)
+            self.env['bus.bus'].sendone(channel, {
+                'type': 'ir.async',
+                'id': job['id'],
+                'name': job['name'],
+                'state': 'processing',
+            })
+            self._cr.commit()
+
+    def _post_process(self, job):
+        # Notify the user a task is completed on the server side
+        super()._post_process(job)
+        if job['notify']:
+            channel = (self._cr.dbname, 'res.partner', self.env.user.partner_id.id)
+            self.env['bus.bus'].sendone(channel, {
+                'type': 'ir.async',
+                'id': job['id'],
+                'name': job['name'],
+                'state': job['state'],
+                'payload': job['payload'],
+            })
+
+    def complete(self):
```

Comment:
why public?

Reply:
JS code calls this endpoint via RPC to mark the result of a SUCCEEDED task has been processed (like the attachment file generated by the asynchronous export has been downloaded). We use this method to mark the task as DONE on all tabs.

Reply:
Then you need to fix the security.

Reply:
How ?

Reply:
`self.check_access_rights('write')` ?

Reply:
There were ir.rules that were allowing the user to access (rwu) the tasks associated to its user but we thought they were useless as everything in the jobqueue is done via pure SQL due to the several locks in postgres.

Reply:
So, is there anything apart of `and user_id = self.env.uid` that can be done ? Shall we reintroduce the ir.rules ?

Reply:
also, beside changing state == "succeded" to state == "done" the user must not edit anything on the record